From 46a077211cc73c15e81eab1230dc462f30dd059c Mon Sep 17 00:00:00 2001
From: Mikhail Paulyshka <me@mixaill.net>
Date: Sun, 18 May 2025 17:52:32 +0300
Subject: [PATCH v2 1/1] x86/rdrand: disable RDSEED on AMD Cyan Skillfish

AMD Cyan Skillfish (Family 17h, Model 47h) has an error that causes
RDSEED to always return 0xffffffff, while RDRAND works correctly.

This patch masks the RDSEED cpu cap for this CPU so that both
/proc/cpuinfo and direct CPUID reads report RDSEED as unavailable.

In addition, it disables visibility of RDRAND and RDSEED through the
MSR on AMD processor.

v2:
  * Drop the MSR writes on Hygon processors.
  * Replace the runtime RDSEED sanity check with a simple
    family/model match for Cyan Skillfish.

Signed-off-by: Mikhail Paulyshka <me@mixaill.net>
---
 arch/x86/include/asm/msr-index.h       |  1 +
 arch/x86/kernel/cpu/rdrand.c           | 12 ++++++++++++
 tools/arch/x86/include/asm/msr-index.h |  1 +
 3 files changed, 14 insertions(+)

diff --git a/arch/x86/include/asm/msr-index.h b/arch/x86/include/asm/msr-index.h
index e7d2f460fcc6..2333f4e7bc2f 100644
--- a/arch/x86/include/asm/msr-index.h
+++ b/arch/x86/include/asm/msr-index.h
@@ -624,6 +624,7 @@
 #define MSR_AMD64_OSVW_STATUS		0xc0010141
 #define MSR_AMD_PPIN_CTL		0xc00102f0
 #define MSR_AMD_PPIN			0xc00102f1
+#define MSR_AMD64_CPUID_FN_7		0xc0011002
 #define MSR_AMD64_CPUID_FN_1		0xc0011004
 #define MSR_AMD64_LS_CFG		0xc0011020
 #define MSR_AMD64_DC_CFG		0xc0011022
diff --git a/arch/x86/kernel/cpu/rdrand.c b/arch/x86/kernel/cpu/rdrand.c
index eeac00d20926..a1a95a956048 100644
--- a/arch/x86/kernel/cpu/rdrand.c
+++ b/arch/x86/kernel/cpu/rdrand.c
@@ -11,6 +11,7 @@
 #include <asm/processor.h>
 #include <asm/archrandom.h>
 #include <asm/sections.h>
+#include <asm/msr.h>
 
 /*
  * RDRAND has Built-In-Self-Test (BIST) that runs on every invocation.
@@ -45,6 +46,17 @@ void x86_init_rdrand(struct cpuinfo_x86 *c)
 	if (failure) {
 		clear_cpu_cap(c, X86_FEATURE_RDRAND);
 		clear_cpu_cap(c, X86_FEATURE_RDSEED);
+		if (c->x86_vendor == X86_VENDOR_AMD) {
+			msr_clear_bit(MSR_AMD64_CPUID_FN_1, 62);
+			msr_clear_bit(MSR_AMD64_CPUID_FN_7, 18);
+		}
 		pr_emerg("RDRAND is not reliable on this platform; disabling.\n");
 	}
+
+	/* disable RDSEED on AMD Cyan Skillfish because of hw/ucode bug */
+	if (c->x86_vendor == X86_VENDOR_AMD && c->x86 == 0x17 && c->x86_model == 0x47) {
+		clear_cpu_cap(c, X86_FEATURE_RDSEED);
+		msr_clear_bit(MSR_AMD64_CPUID_FN_7, 18);
+		pr_emerg("RDSEED is not reliable on this platform; disabling.\n");
+	}
 }
diff --git a/tools/arch/x86/include/asm/msr-index.h b/tools/arch/x86/include/asm/msr-index.h
index e6134ef2263d..8b48a54b627a 100644
--- a/tools/arch/x86/include/asm/msr-index.h
+++ b/tools/arch/x86/include/asm/msr-index.h
@@ -616,6 +616,7 @@
 #define MSR_AMD64_OSVW_STATUS		0xc0010141
 #define MSR_AMD_PPIN_CTL		0xc00102f0
 #define MSR_AMD_PPIN			0xc00102f1
+#define MSR_AMD64_CPUID_FN_7		0xc0011002
 #define MSR_AMD64_CPUID_FN_1		0xc0011004
 #define MSR_AMD64_LS_CFG		0xc0011020
 #define MSR_AMD64_DC_CFG		0xc0011022
-- 
2.49.0

